### PR TITLE
Feat/serializer context decoupling

### PR DESCRIPTION
**Pull request by maurei**
# Overview decoupling PR
This is an intermediate PR intended to provide insight into the structural decisions made so far. Please limit feedback to those elements. Commenting, variable naming and spacing consistency will be further improved along the way.

This PR contains rework of three main elements:
- Deserializer 
- Serializer 
- Breaking up of RequestManager into more separated isolated services [WIP]
- Breaking up ResourceGraph [WIP]

# Serializers and Deserializers

see wiki [here](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/feat/serializer-context-decoupling/src/JsonApiDotNetCore/Serialization/wiki.md).

# RequestManager
The `RequestManager` was becoming too big and it contained too many responsibilities. I have introduced the following approach.
The request manager is split up into services in one of two categories:
- URL query param services: stuff related to anything behind the `?` in URL, like inclusions, filters, etc.
- request scope meta services: remaining meta data associated to scope of one request, like the main resource of the request (eg. `Article` when route like `/articles**`).

#### URL query param services
Each URL query operator has been moved to its own dedicated service. This is to prevent that if a service needs only a small subset (eg IncludedRelationshipBuilder needs only the inclusion chains), it does not exposed to all the other query parameters too.
 * `IPageQueryService` the former page manager
 * `IFieldQueryService` to access sparse field selections
 * `IIncludedQueryService` to access inclusion chains
 * `IAttributeBehaviourQueryService` to encapsulate the feature described [here](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/master/src/JsonApiDotNetCore/Configuration/NullAttributeResponseBehavior.cs)
 * Others are yet to be implemented: `FilterQueryService`, `SortQueryService`. I might be missing more.

Each service will also contain "business logic" related to that particular service. For example IIncludedQueryService
- knows how to convert a chain ("articles.author.favorite-food") into corresponding `RelationshipAttribute`s 
- throws errors when a non-existing relationship is targeted

#### Request scope meta services
Remaining metadata associated to a particular request. Also separated into multiple small services to prevent unnecessary exposure of data.
* `IUpdatedFields` to keep track of which fields of a resource (`AttrAttribute`s and `RelationshipAttribute`s) are targeted in the scope of a request.
* `ICurrentRequest` to access metadata that is not (yet) isolated to its own service. Like the main resource of a request.

#### Todo:
see the *App vs. Request vs. Operations* section in #253. Currently the RequestManager had both request-level and operational-level  state information (updated attributes/relationships vs main request resource).
This particular example has been separated but we need to further categorize the above services such that it is compatible with this distinction.

We will currently not support bulk operations (dedicated issue + PR coming soon) for v4, but in order to re-support it in the future it is important to properly distinguish between operation-scoped services and request-scoped services (eg. `IUpdatedFields` is operation-level, `ICurrentRequest` request-level). All requests will then, for now, be associated to one operation, but can be associated with more.

# ResourceGraph
Throughout JADNC, the resource graph is used for two purposed
- getting `ContextEntity` given a exposed resource string, resource type, controller name
- resolving relationships which require a lot of information from the graph, eg many-to-many relationships.

Because of these two pretty deviating responsibilities, I have introduced
- `IContextEntityProvider` just to get a `ContextEntity` (various overloads)
- still need to put the rest of resource graph behind a new interface, something like `IRelationshipResolver`, which can probably be merged with the already existing `IInverseRelationshipResolver` which was introduced in `ResourceHooks`.

# Todo
Few things are left to be done in the domain of serialization:

- Support `camelCase` configuration as member names and resource names. Need to think about how to make this configurable in such a way that it #555 is addressed.
- Sparse field selection on included members in `IncludedRelationshipBuilder`.

Things left to be done in general
- Wiring back up parsing of URL query parameters in `QueryParser`
- Applying these queries in the data layer: `JsonaApiContext` is still in there, we need to create something like a `QueryProcessorFactory` service that instantiates eg `SortQuery`s or `FilterQuery`s using DI, and use that in the places were manual instantiation is being performed right now (see the DefaultEntityRepository class).
- Removing the operations experimental feature. We might put it back later but for now it would delay v4 release by too much time. Also the current implementation deviates a lot from the extension that might be included in the v1.1 specs
- Removing the EntityResource separation example and related code. See #553

### CODE REVIEW

**Comment by wisepotato** [comment]
image
 
ugh.

File: src/JsonApiDotNetCore/Services/QueryAccessor.cs [hunk 1]
```diff
@@ -32,7 +32,7 @@ public class QueryAccessor : IQueryAccessor
         /// <param name="requestManager"></param>
         /// <param name="logger"></param>
         public QueryAccessor(
-            IRequestManager requestManager,
+            ICurrentRequest requestManager,
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Services/QueryAccessor.cs]: variable

File: src/JsonApiDotNetCore/Services/QueryComposer.cs [hunk 2]
```diff
@@ -6,12 +6,12 @@ namespace JsonApiDotNetCore.Services
 {
     public interface IQueryComposer
     {
-        string Compose(IRequestManager jsonApiContext);
+        string Compose(ICurrentRequest jsonApiContext);
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Services/QueryComposer.cs]: variable

File: src/JsonApiDotNetCore/Services/QueryComposer.cs [hunk 3]
```diff
@@ -6,12 +6,12 @@ namespace JsonApiDotNetCore.Services
...
     }
 
     public class QueryComposer : IQueryComposer
     {
-        public string Compose(IRequestManager requestManager)
+        public string Compose(ICurrentRequest requestManager)
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Services/QueryComposer.cs]: variable

File: src/JsonApiDotNetCore/Services/QueryParser.cs [hunk 4]
```diff
@@ -4,28 +4,44 @@
...
+using JsonApiDotNetCore.Internal.Contracts;
 using JsonApiDotNetCore.Internal.Query;
 using JsonApiDotNetCore.Managers.Contracts;
 using JsonApiDotNetCore.Models;
+using JsonApiDotNetCore.QueryServices.Contracts;
 using Microsoft.AspNetCore.Http;
 
 namespace JsonApiDotNetCore.Services
 {
+
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Services/QueryParser.cs]: cheeky! but not needed

File: src/JsonApiDotNetCore/QueryServices/Contracts/IFieldQueryService.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+using JsonApiDotNetCore.Models;
+
+namespace JsonApiDotNetCore.QueryServices.Contracts
+{
+    public interface IFieldsQueryService
+    {
+        List<AttrAttribute> Get(RelationshipAttribute relationship = null);
+    }
+
+    public interface IInternalFieldsQueryService
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/Contracts/IFieldQueryService.cs]: commentssss and class per file

File: src/JsonApiDotNetCore/QueryServices/Contracts/IIncludedQueryService.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using JsonApiDotNetCore.Models;
+
+namespace JsonApiDotNetCore.QueryServices.Contracts
+{
+    public interface IIncludedQueryService
+    {
+        List<List<RelationshipAttribute>> Get();
+    }
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/Contracts/IIncludedQueryService.cs]: comments + class per file

File: src/JsonApiDotNetCore/QueryServices/Contracts/IPageQueryService.cs [hunk 7]
```diff
@@ -5,9 +5,9 @@
 using System.Collections.Generic;
 using System.Text;
 
-namespace JsonApiDotNetCore.Managers.Contracts
+namespace JsonApiDotNetCore.QueryServices.Contracts
 {
-    public interface IPageManager
+    public interface IPageQueryService
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/Contracts/IPageQueryService.cs]: comment what i could do with this, its public

File: src/JsonApiDotNetCore/QueryServices/Contracts/IPageQueryService.cs [hunk 8]
```diff
@@ -29,7 +29,8 @@ public interface IPageManager
         /// Are we even paginating
         /// </summary>
         bool IsPaginated { get; }
+        int TotalPages { get; }
 
-        RootLinks GetPageLinks();
+        bool ShouldPaginate();
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/Contracts/IPageQueryService.cs]: difference betwen this and `ShouldPaginate`?

Comment (wisepotato) [reply src/JsonApiDotNetCore/QueryServices/Contracts/IPageQueryService.cs]: also all the other methods are commented, not these

File: src/JsonApiDotNetCore/QueryServices/FieldQueryService.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
+﻿using System.Collections.Generic;
+using JsonApiDotNetCore.Models;
+using JsonApiDotNetCore.QueryServices.Contracts;
+
+namespace JsonApiDotNetCore.QueryServices
+{
+    public class FieldsQueryService : IFieldsQueryService, IInternalFieldsQueryService
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/FieldQueryService.cs]: ye olde comments plz

Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/FieldQueryService.cs]: I wanted to check what this does by means of tests, should be unit tested

File: test/UnitTests/Services/QueryParserTests.cs [hunk 10]
```diff
@@ -40,7 +44,7 @@ public void Can_Build_Filters()
                 .Setup(m => m.DisabledQueryParams)
                 .Returns(QueryParams.None);
 
-            var queryParser = new QueryParser(_requestMock.Object, new JsonApiOptions());
+            var queryParser = new QueryParser(_includedQuery, _fieldsQuery, _requestMock.Object, _graph, new JsonApiOptions());
```
Comment (wisepotato) [inline test/UnitTests/Services/QueryParserTests.cs]: i see this coming back 5x, can we abstract this away?

File: test/UnitTests/Services/QueryAccessorTests.cs [hunk 11]
```diff
@@ -13,13 +13,13 @@ namespace UnitTests.Services
...
     {
-        private readonly Mock<IRequestManager> _rmMock;
+        private readonly Mock<ICurrentRequest> _rmMock;
         private readonly Mock<ILogger<QueryAccessor>> _loggerMock;
         private readonly Mock<IQueryCollection> _queryMock;
 
         public QueryAccessorTests()
         {
-            _rmMock = new Mock<IRequestManager>();
+            _rmMock = new Mock<ICurrentRequest>();
```
Comment (wisepotato) [inline test/UnitTests/Services/QueryAccessorTests.cs]: `_crMock`, i know its annoying but we need to be consistent

File: src/JsonApiDotNetCore/QueryServices/IncludedQueryService.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Collections.Generic;
+using JsonApiDotNetCore.Models;
+using JsonApiDotNetCore.QueryServices.Contracts;
+
+namespace JsonApiDotNetCore.QueryServices
+{
+
+    public class IncludedQueryService : IIncludedQueryService, IInternalIncludedQueryService
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/IncludedQueryService.cs]: comments

File: src/JsonApiDotNetCore/QueryServices/PageQueryService.cs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        private IJsonApiOptions _options;
+
+        public PageQueryService(IJsonApiOptions options)
+        {
+            _options = options;
+            DefaultPageSize = _options.DefaultPageSize;
+            PageSize = _options.DefaultPageSize;
+        }
+        public int? TotalRecords { get; set; }
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryServices/PageQueryService.cs]: properties/fields before constructor

File: src/JsonApiDotNetCore/RequestServices/CurrentRequest.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+using JsonApiDotNetCore.Services;
+using Microsoft.AspNetCore.Http;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace JsonApiDotNetCore.Managers
+{
+
+    class CurrentRequest : ICurrentRequest
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/RequestServices/CurrentRequest.cs]: comments + very nice, clean small class

File: src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+            if (field is AttrAttribute)
+                return;
+
+            // can't provide any more data other than the rios since it is not contained in the included section
+            if (_document.Included == null || _document.Included.Count == 0)
+                return;
+
+            if (field is HasOneAttribute hasOneAttr)
+            {
+                var rio = data.SingleData;
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs]: rio?

File: src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+        private ResourceObject GetLinkedResource(ResourceIdentifierObject relatedResourceIdentifier)
+        {
+            try
+            {
+                return _document.Included.SingleOrDefault(r => r.Type == relatedResourceIdentifier.Type && r.Id == relatedResourceIdentifier.Id);
+            }
+            catch (InvalidOperationException e)
+            {
+                throw new InvalidOperationException($"A compound document MUST NOT include more than one resource object for each type and id pair."
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs]: any explanation on what exactly is happening here? What is a compound document

File: src/JsonApiDotNetCore/Builders/IDocumentBuilder.cs [hunk 17]
```diff
@@ -32,6 +31,6 @@ public interface IDocumentBuilder
         /// that should not be exposed to the client. For example, you might want to limit
         /// the exposed attributes based on the authenticated user's role.
         /// </param>
-        ResourceObject GetData(ContextEntity contextEntity, IIdentifiable entity, IResourceDefinition resourceDefinition = null);
+        ResourceObject GetData(ContextEntity contextEntity, IIdentifiable entity,  object resourceDefinition = null);
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Builders/IDocumentBuilder.cs]: are you absolutely sure? `object`?

File: src/JsonApiDotNetCore/Services/Operations/Processors/CreateOpProcessor.cs [hunk 18]
```diff
@@ -67,7 +67,8 @@ public async Task<Operation> ProcessAsync(Operation operation)
             // can locate the result of this operation by its localId
             operationResult.DataObject.LocalId = operation.DataObject.LocalId;
 
-            return operationResult;
+            return null;
```
[the remark this answers is not in the document]
Comment (maurei) [reply src/JsonApiDotNetCore/Services/Operations/Processors/CreateOpProcessor.cs]: yes - intermediate PR.  Operations is to be removed in separate PR

File: src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs (hunk 15 above)
Comment (maurei) [reply src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs]: rio = resource identifier object
see technical language used in json:api specs

File: src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs (hunk 16 above)
Comment (maurei) [reply src/JsonApiDotNetCore/Serialization/Deserializer/ClientDeserializer.cs]: we're adhering to json:api technical language throughout the framework

https://jsonapi.org/format/#document-compound-documents

**Pull request merged**